src/references.md

```diff
@@ -17,161 +17,140 @@ Which obey the following rules:
+Specifically, we must concern ourselves with the existence of this degenerate
+input: `compute(&x, &mut x)`. With this input, we would get this execution:
 
-# Liveness
+```rust
+                    // *input == *output == 20 (just some large value)
+if *input > 10 {    // true  (input == 20)
+    *output = 1;
+}
+if *input > 5 {     // false (input == 1)
+    *output *= 2;
+}
+                    // *input == *output == 1
+```
 
-Note: Liveness is not the same thing as a *lifetime*, which will be explained
-in detail in the next section of this chapter.
+which our optimized function would produce `output == 2` for. In Rust we
+know this input should be impossible because `&mut` isn't allowed to be
+aliased, and so we can safely reject its possibility.
 
-Roughly, a reference is *live* at some point in a program if it can be
-dereferenced. Shared references are always live unless they are literally
-unreachable (for instance, they reside in freed or leaked memory). Mutable
-references can be reachable but *not* live through the process of *reborrowing*.
+Alias analysis is important because it lets the compiler perform several
+useful optimizations, such as:
 
-A mutable reference can be reborrowed to either a shared or mutable reference to
-one of its descendants. A reborrowed reference will only be live again once all
-reborrows derived from it expire. For instance, a mutable reference can be
-reborrowed to point to a field of its referent:
+* keeping current values in hardware registers as long as possible by proving no pointer touches it
+* eliminating reads by proving the value hasn't been written too since last we read
+* eliminating writes by proving that the value is never read before the next write
+* moving or reordering reads and writes by proving they don't depend on each other
 
-```rust
-let x = &mut (1, 2);
-{
-    // reborrow x to a subfield
-    let y = &mut x.0;
-    // y is now live, but x isn't
-    *y = 3;
-}
-// y goes out of scope, so x is live again
-*x = (5, 7);
-```
+In the previous example, we used the fact that `&mut u32` can't alias to prove
+that writes to `output` can't possibly affect `input`. This let us cache `input`
+in a register, eliminating a read. By caching this read, we knew that the the
+write in the `> 10` branch couldn't affect whether we take the `> 5` branch,
+allowing us to also eliminate a read-modify-write (doubling the output) when
+`input > 10`.
 
-It is also possible to reborrow into *multiple* mutable references, as long as
-they are *disjoint*: no reference is an ancestor of another. Rust
-explicitly enables this to be done with disjoint struct fields, because
-disjointness can be statically proven:
+The key thing to know about aliasing is that writes are the only thing that
+really matter. For instance, we have little concern for aliasing in this
+modified version of our function, because we've moved the writes to a temporary
+variable:
```

Comment:
That doesn't sound entirely right to me. Consider the following:
```rust
fn foo(x: &i32) -> i32 {
  let y = *x;
  some_unknown_function();
  return y;
}
```
It should be possible to re-order the load and the function call, without knowing anything about that function (other than it being safe).
Now, one may argue that ultimately, this is about whether the unknown code writes to `x` or not. Still, saying that reads "don't really matter" seems to be an overstatement to me.